### PR TITLE
Workflow management API breaking changes for v1.11

### LINKED ISSUE #5911: [Workflow] Improvements to existing workflow management APIs
## In what area(s)?

> /area runtime

## Describe the feature

The initial release of the Workflow management API in v1.10 included APIs for starting workflows, querying their status, and terminating workflows. However, there are some problems with the implementation that can make them a bit hard to use.

### Start Workflow

The URL template is as follows: `POST workflows/{workflowComponent}/{workflowType}/{instanceId}/start`

#### Proposed improvements
* **Make the instance ID optional**  
Today it's required, but we shouldn't force users to come up with instance IDs. If one isn't provided, we should let the component generate a random ID. This will likely require a new HTTP route.

* **Return `Location` and `Retry-After` response headers from the HTTP handler**  
Today we return a HTTP 202 response but without a `Location` header. We should add this header to the response and have it point to the GetWorkflowStatus API. The goal is to natively support the [asynchronous request-reply pattern](https://learn.microsoft.com/azure/architecture/patterns/async-request-reply) out of the box when invoking workflows. The `Retry-After` header can be mainly to ensure that consumers of this pattern do not try to poll too frequently.

* **Return more information in the response**  
The current response payload is very sparse: `{"instance_id":"..."}` and doesn't follow our existing JSON field casing conventions. We should enrich this with more information and use consistent casing for JSON fields. Here's an example:

```json
{
  "workflow": {
    "instanceID": "{instanceID}"
  },
  "httpManagement": {
    "status": {"method":"GET","url":"http://localhost:3500/v1.0-alpha1/workflows/{componentName}/{instanceID}"},
    "terminate": {"method":"POST","url":"http://localhost:3500/v1.0-alpha1/workflows/{componentName}/{instanceID}"},
    "suspend": {"method":"POST","url":"http://localhost:3500/v1.0-alpha1/workflows/{componentName}/{instanceID}/suspend"},
    "resume": {"method":"POST","url":"http://localhost:3500/v1.0-alpha1/workflows/{componentName}/{instanceID}/resume"}
  }
}
```  

  Note that `{componentName}` and `{instanceID}` would be filled in with the relevant values. Also, the HTTP scheme and hostname will match the schema and hostname of the incoming request (with overrides allowed by using `X-Forwarded-*` request headers). Users (or code) should be able to simply use the URLs as-is to manage the created workflow instance.

* **Don't parse the input in the implementation**
Today we parse the input in the API handler, assuming that the payload is JSON. This API should actually just pass-through the payload to the component without trying to interpret it. Otherwise, we're adding unnecessary performance overhead and restricting all implementations to only support JSON.

### Get Workflow

The URL template is as follows: `GET workflows/{workflowComponent}/{workflowType}/{instanceId}`

#### Proposed improvements

* **Remove the `{workflowType]` segment from the URL**  
This segment is not needed. The only thing that's needed is {workflowComponent} and {instanceID}.

* **Rename "WFInfo" to "workflow"**  
This is to 1) make the payload similar to the proposed "Start Workflow" response payload and 2) follow are more conventional model for JSON payloads.

* **Change timestamp** fields in protobuf contract from string/int to `google.protobuf.Timestamp`**
This is a more proper way to represent timestamp information and will help avoid compatibility issues between SDKs and components.

* **Surface more metadata to the top-level `"workflow"` JSON object**  
Today the response payload returns an instance ID, a created timestamp, and a bag of metadata values. Many of these metadata values should be promoted to be 1st-class fields in the `workflow` object. This will make the JSON easier to parse and deserialize into rich types by the SDKs.  Here's an example:

```json
{
  "workflow": {
    "instanceID": "order123",
    "createdAt": "2023-02-02T07:00:00Z",
    "lastUpdatedAt": "2023-02-02T12:00:00Z",
    "runtimeStatus": "RUNNING"
  }
}
```

As an aside, the internal implementation should remove the `WorkflowReference` struct and just use `instanceID` (string) to identify workflow instances.

## Release Note

RELEASE NOTE: **UPDATE** workflow management APIs

/cc @nyemade-uversky

### LINKED ISSUE #6156: Workflows: actor ID should not use the `#` character (Workflows not working with Cosmos DB)
## In what area(s)?

/area runtime

## What version of Dapr?

> 1.10.x
> master

## Description

This issue was originally reported by @mPyKen here:

When Workflows creates the actor ID, it includes a `#` in the name: https://github.com/dapr/components-contrib/issues/2071#issuecomment-1480914079

https://github.com/dapr/dapr/blob/beef33d04f9528e312fe5d71ab6318a9c0728830/pkg/runtime/wfengine/workflow.go#L466-L470

This makes Workflows not usable with certain state stores such as Cosmos DB.

After discussing with @cgillum offline, we think the best approach may be to change how actor IDs are named by workflows, using a different character. This will be a breaking change (on an alpha building block), but because the `#` is not commonly used, we are unsure what other issues may appear.

### PR DESCRIPTION
**Pull request by cgillum**
# Description

This PR contains a set of breaking changes to improve the usability of the workflow management APIs. These changes are expected to go out with Dapr v1.11.

**Corresponding components-contrib PR: https://github.com/dapr/components-contrib/pull/2751**

## Changes

There are several changes in this PR, many of them breaking.

### Start Workflow API
* **[Breaking from v1.10]** Removed `{instanceID}` from the path and made it an optional query string parameter.
* **[Breaking from v1.10]** Changed response type and format
* **[Breaking from v1.10]** Instance IDs must now only have letters, numbers, dashes, and underscores and be < 64 characters
* **[Breaking from v1.10]** Workflow input now comes from raw HTTP request payload - no parsing
* If instance ID is not specified, a random GUID is chosen

### Get Workflow API
* **[Breaking from v1.10]** Changed format of JSON response
* **[Breaking from v1.10]** Moved several "metadata" fields to be first-class in the JSON response to make it more usable

### Raise Event API
* Event payload now comes from raw HTTP request payload - no parsing

### Terminate / Pause / Resume Workflow APIs (new in v1.11)
* Changed gRPC contracts to be distinct (previously they all shared the same contract)

### Test updates
* Minor refactoring / simplification
* Added validation of HTTP response payloads
* Workflow API tests now have 99% code coverage

### Misc. updates
* Added additional clarity in API error messages

## Issue reference

Resolves #5911
Resolves #6156

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [x] Created/updated tests
* [x] Unit tests passing
* [x] End-to-end tests passing
* [x] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#3352
* [x] Specification has been updated / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#3352
* [ ] ~Provided sample for the feature / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_~

### CODE REVIEW

File: dapr/proto/runtime/v1/dapr.proto
Comment (DeepanshuA) [inline dapr/proto/runtime/v1/dapr.proto:916]: As the attributes for Terminate, Pause and Resume Request seems to be same, should we not keep a common request?

Comment (cgillum) [reply dapr/proto/runtime/v1/dapr.proto:916]: No, I want to better future proof these APIs by giving each their own request message. That way we can evolve them independently over time.

**Comment by cgillum** [comment]
NOTE: this PR is currently blocked waiting for https://github.com/dapr/dapr/pull/6186 to be merged.

**Comment by cgillum** [comment]
This PR is unblocked now and should be good to go. :)  Would appreciate if an approver/maintainer could approve the PR validation workflows to run, and provide any PR feedback.

File: dapr/proto/runtime/v1/dapr.proto
Comment (mukundansundar) [reply dapr/proto/runtime/v1/dapr.proto:888]: Does this specifically have to be timestamp? 
cc @dapr/maintainers-dapr

Comment (cgillum) [reply dapr/proto/runtime/v1/dapr.proto:888]: Timestamp is the natural choice. Choosing `string` creates a maintainability problem since you rely on both sides of every timestamp field to agree on a particular format (e.g. RFC3339). It's also more efficient over the wire. I recommend the Dapr project use timestamps instead of strings in new protobuf contracts moving forward.

Comment (ItalyPaleAle) [reply dapr/proto/runtime/v1/dapr.proto:888]: I second @cgillum. If not `google.protobuf.Timestamp`, then perhaps a uint64 with a UNIX timestamp as another option.

Comment (cgillum) [reply dapr/proto/runtime/v1/dapr.proto]: I went ahead and updated the protobuf field name to be `created_at` to match the JSON name. I also renamed  `last_updated_time` to `last_updated_at` to match the corresponding JSON field (apologies that this change got included in my commit to merge latest from master).

File: pkg/grpc/universalapi/api_workflow.go
Comment (DeepanshuA) [inline pkg/grpc/universalapi/api_workflow.go:222]: Do we want to convert letters to all upper/lower - so that instance ids are not just differentiated in some cases on the basis of case sensitivity?

File: go.mod
Comment (DeepanshuA) [inline go.mod:430]: I know we have a code comment there, but just commenting here explicitly, so that we remember to update it before merging this PR.

File: pkg/grpc/universalapi/api_workflow.go
Comment (cgillum) [reply pkg/grpc/universalapi/api_workflow.go:222]: I like the idea of having workflow IDs be case-insensitive but I'm not sure if I'd want to enable that through a transformation in the API layer. I would prefer something like that be done at the engine layer - probably not part of this PR.

Comment (ItalyPaleAle) [inline pkg/grpc/universalapi/api_workflow.go:222]: Does this need to support non-ASCII characters? I am not sure if that's a risk. The current code iterates through runes, and methods like `unicode.IsLetter` match the entire "letter" plane in the Unicode specs and not just ASCII letters.

```go
	var chr byte
// Must not use `for .. range` because that iterates through runes (i.e. UTF-8 characters), while a regular for loop iterates through bytes
	for i := 0; i < len(instanceID); i++ {
		chr = instanceID[i]
		if (chr < 'A' || chr > 'Z') && (chr < 'a' || chr > 'z') && (chr < '0' || chr > '9') && chr != '_' && chr != '-' {
			return messages.ErrInvalidInstanceID.WithFormat(instanceID)
		}
	}
```

File: pkg/http/api.go
Comment (ItalyPaleAle) [inline pkg/http/api.go:697]: Should this be passed in the body?

File: pkg/grpc/universalapi/api_workflow.go
Comment (ItalyPaleAle) [inline pkg/grpc/universalapi/api_workflow.go:100]: Nit: this can cause an unnecessary heap allocation (and so more work for the GC) when it's not needed. I'd recommend:

```
return &emptypb.Empty{}, err
```

File: pkg/grpc/universalapi/api_workflow.go
Comment (ItalyPaleAle) [inline pkg/grpc/universalapi/api_workflow.go:214]:
```suggestion
		const maxInstanceIdLength = 64
```

File: pkg/http/api.go
Comment (ItalyPaleAle) [inline pkg/http/api.go:728]: Why was this undone?

File: pkg/runtime/wfengine/component.go
Comment (ItalyPaleAle) [inline pkg/runtime/wfengine/component.go:172]: Can these be constants?

File: pkg/grpc/universalapi/api_workflow.go
Comment (cgillum) [reply pkg/grpc/universalapi/api_workflow.go:222]: There's no requirement to support non-ASCII characters. I figured we can start out as being a bit strict and then loosen up the requirements over time based on user feedback.

File: pkg/http/api.go
Comment (cgillum) [reply pkg/http/api.go:697]: No, I'm intentionally trying to use the request body exclusively for the raw workflow input without any Dapr interpretation.

File: pkg/grpc/universalapi/api_workflow.go
Comment (cgillum) [reply pkg/grpc/universalapi/api_workflow.go:100]: There shouldn't be any unnecessary allocations because every code path returns the same `emptyResponse` value, whether there is an error or a success. Or am I misunderstanding?

File: pkg/http/api.go
Comment (cgillum) [reply pkg/http/api.go:728]: Yes, this was removed intentionally. I wanted to make the terminate, pause, and resume APIs more future-proof by giving them their own distinct contract and implementation, which means we don't use a common API handler for all of them.

File: pkg/runtime/wfengine/component.go
Comment (cgillum) [reply pkg/runtime/wfengine/component.go:172]: They can, but is that what we want when they're only used in one place? I didn't see an advantage to using constants for these, only disadvantages.

File: pkg/http/api.go
Comment (ItalyPaleAle) [reply pkg/http/api.go:697]: Ok the same situation was just addressed here: https://github.com/dapr/dapr/pull/6197#discussion_r1170857312

@mukundansundar  and @artursouza recommended to use HTTP headers instead

File: pkg/grpc/universalapi/api_workflow.go
Comment (ItalyPaleAle) [reply pkg/grpc/universalapi/api_workflow.go:100]: Oh I'm sorry, I missed that it was returned also in case of success.

File: pkg/runtime/wfengine/component.go
Comment (ItalyPaleAle) [reply pkg/runtime/wfengine/component.go:172]: The advantage would be if they were used in unit tests too. Otherwise, ok

File: pkg/grpc/universalapi/api_workflow.go
Comment (ItalyPaleAle) [reply pkg/grpc/universalapi/api_workflow.go:222]: I agree. Can you change to use the loop as in my example above then? The current implementation does support unicode broadly

File: pkg/http/api.go
Comment (cgillum) [reply pkg/http/api.go:697]: That's interesting. I'd love to get more context on this statement, which I don't understand:
> My concern is that query args are directly available to be cached and used.

Happy to add support for headers. Is supporting both an option? I do like the convenience of query string parameters.

File: pkg/runtime/wfengine/component.go
Comment (cgillum) [reply pkg/runtime/wfengine/component.go:172]: My counter to that is duplicating these values in unit tests helps prevent accidental breaking changes (a unit test is a bit more representative of customer code this way). :)

File: pkg/grpc/universalapi/api_workflow.go
Comment (cgillum) [reply pkg/grpc/universalapi/api_workflow.go:222]: Oh, I missed that detail. That's interesting because I actually started by borrowing logic from [token.IsIdentifier](https://github.com/golang/go/blob/598cf5e6ac1deb87f4931af195dad66f2b3f1672/src/go/token/token.go#L335) for this code. I'm not an expert on this topic by any means, but any idea why the go maintainers were okay with using `range` on a string instead of treating it as an array like you're suggesting?

File: pkg/http/api.go
Comment (DeepanshuA) [reply pkg/http/api.go:698]: @mukundansundar @artursouza What is your viewpoint here?

File: go.mod
Comment (cgillum) [reply go.mod]: This has been removed. We're now getting the latest from `dapr/components-contrib`.

**Comment by cgillum** [comment]
@ItalyPaleAle @artursouza @mukundansundar there are a couple threads still unresolved, but this PR is otherwise ready to go.

Unfortunately, it looks like GitHub is having a partial outage, which is causing some CI failures. Independent of that, I'd love to get a sign-off. The corresponding components-contrib PR has been merged: https://github.com/dapr/components-contrib/pull/2751.

File: pkg/grpc/universalapi/api_workflow.go
Comment (cgillum) [reply pkg/grpc/universalapi/api_workflow.go:222]: Resolving for now as I don't think this should be a blocker. I'd love to hear your opinion on this when you get back from vacation, though!

**Comment by cgillum** [comment]
The two failing tests in the most recent run appear to be flakey unit tests related to actors. They aren't related in any way to this PR and can likely be scheduled for retry by a maintainer. This PR is otherwise good to be merged, pending any blocking PR comments.

File: pkg/grpc/universalapi/api_workflow.go
Comment (mukundansundar) [inline pkg/grpc/universalapi/api_workflow.go:145]: should this be changed to `EventData` in the incoming request as well?

File: pkg/grpc/universalapi/api_workflow.go
Comment (mukundansundar) [inline pkg/grpc/universalapi/api_workflow.go:172]: would prefer consistency in casing.

File: pkg/grpc/universalapi/api_workflow.go
Comment (mukundansundar) [inline pkg/grpc/universalapi/api_workflow.go:230]:
```suggestion
func (a *UniversalAPI) getWorkflowComponent(componentName string) (workflows.Workflow, error) {
```
There can be other methods with similar names later on for other components.

File: pkg/grpc/universalapi/api_workflow.go
Comment (mukundansundar) [inline pkg/grpc/universalapi/api_workflow.go:214]: Will the instance Id be prefixed with `dapr`? So that it is known that it is dapr specific.

File: pkg/http/api.go
Comment (mukundansundar) [reply pkg/http/api.go:698]: In this case I agree that queryArgs will be better, as this is specific for an instance and optional.

File: pkg/grpc/universalapi/api_workflow.go
Comment (cgillum) [reply pkg/grpc/universalapi/api_workflow.go:51]: A few things about this:

- Dapr has linting rules that flag "Id" and require us to change to "ID".
- "Id" is a .NET convention whereas all other languages use "ID".
- The `InstanceId` in this line comes from the durabletask gRPC proto definitions, which are external to Dapr and can't be changed right now.
- The `InstanceID` in the workflow struct comes from components-contrib, which has already been merged for 1.11. Changing this now would require another round of changes in components-contrib.

With this context, do you still feel strongly that we should change this now?

File: pkg/grpc/universalapi/api_workflow.go
Comment (cgillum) [reply pkg/grpc/universalapi/api_workflow.go:172]: See my previous response on casing.

File: pkg/grpc/universalapi/api_workflow.go
Comment (cgillum) [reply pkg/grpc/universalapi/api_workflow.go:214]: No, these IDs are provided by the user. We don't apply any transformations to them.

File: pkg/grpc/universalapi/api_workflow.go
Comment (cgillum) [reply pkg/grpc/universalapi/api_workflow.go]: Done.

File: pkg/grpc/universalapi/api_workflow.go
Comment (cgillum) [reply pkg/grpc/universalapi/api_workflow.go]: Good call. Done.

File: pkg/grpc/universalapi/api_workflow.go
Comment (artursouza) [reply pkg/grpc/universalapi/api_workflow.go:51]: Everything inside Dapr should be ID. So I am fine with external things being as-is.

**Review by artursouza — APPROVED** [review]

**Pull request merged**
